### PR TITLE
VTGate V3: Feature summary

### PR DESCRIPTION
**Pull request by sougou**
@AndyDiamondstein @michael-berlin @erzel
I've renamed the old vtgatev3 doc to VTGateV3Design, which is not very interesting to someone who wants to use V3. This new doc fulfills the purpose.

### CODE REVIEW

**Comment by AndyDiamondstein** [comment]
It might be more clear to name the new document VTGateV3Features.md or VTGateV3Overview.md or something else that will distinguish it from other VTGateV3... docs down the line (implementation, reference, codelab, etc.).

**Comment by sougou** [comment]
Makes sense. Renamed.

File: doc/VTGateV3Features.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+The order table would also need an order_id column. As mentioned above, you can create an order_id->customer_id cross-shard index for this table. This will allow you to efficiently access orders by their order_id.
+
+In the case of an order_detail table, it may only need an order_id foreign key. Since this foreign key means the same thing as the order_id in order, creating a cross-shard index for it will result in a duplication of the order_id->customer_id index. In such situations, V3 allows you to just reuse the existing index for the order_detail table also. This saves disk space and also reduces the overall write load.
+
+### Consistency
+Once you add multiple indexes to tables, it's possible that the application could make inconstent requests. V3 makes sure that none of the specified constraints are broken. For example, if a table had both a basic sharding key and a hashed sharding key, it will enforce the rule that the hash of the basic sharding key matches that of the hashed sharding key.
```
Comment (erzel) [inline doc/VTGateV3Features.md]: typo: inconstent

File: doc/VTGateV3Features.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+#### Hashed sharding key
+If the application's sharding key is a monotonically increasing number, then you may not get well-balanced shards. In such cases, you can ask V3 to route queries based on the hash of the main sharding key.
+
+Vitess's filtered replication currently requires that the hash value be physically present as a column in each table. To satisfy this need, you still need to create a column to store this hash value. However, V3 will take care of populating this on your behalf. *We will soon be removing this restriction once we change filtered replication to also perform the same hashing.*
+
+#### Auto-increment columns
+When a table gets sharded, you are no longer able to use MySQL's auto increment functionality. V3 allows you to designate a table in an unsharded database as the source of auto-increment ids. Once you've specified this, V3 will transparently use generated values from this table to keep the auto-increment going. Additionally, this column can be a hashed sharding key. This means that the insert will also get routed to the correct shard based on the hashed routing value.
```
Comment (erzel) [inline doc/VTGateV3Features.md]: “Additionally, this column can be a hashed sharding key”. Did you mean the source of a hashed sharding key; i.e. the hash function is operated on this auto-incremented column, right? If so, can you clarify the language.

File: doc/VTGateV3Features.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+### Consistency
+Once you add multiple indexes to tables, it's possible that the application could make inconstent requests. V3 makes sure that none of the specified constraints are broken. For example, if a table had both a basic sharding key and a hashed sharding key, it will enforce the rule that the hash of the basic sharding key matches that of the hashed sharding key.
+
+Some of the changes require updates to be performed across multiple databases. For example, inserting a row into a table that has a cross-shard key requires an additional row to be inserted into the lookup table. This results in distributed transactions. Currently, this is a best effort update. It is possible that partial commits happen if databases fail in the middle of a distributed commit. *We will soon develop support for 2PC to overcome this limitation.*
+
+### Query diversity
+V3 does not support the full SQL feature set. The current implementation supports simple queries:
+* Single table DML statements: This is a vitess-wide restriction where you can affect only one table and one sharding key per statement. *This restriction may be limited in the future.*
```
Comment (erzel) [inline doc/VTGateV3Features.md]: “limited in the future”. You probably mean “lifted in the future”.

File: doc/VTGateV3Features.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+#### Cross-shard indexes
+As your application evolves, you'll invariably find yourself wanting to fetch rows based on columns other than the main sharding key. For example, if you've sharded your database by user id, you may still want to be able find users by their username. If you only had the sharding key, such queries can only be answered by sending it to all shards. This could become very expensive as the number of shards grow.
+
+The typical strategy to address this problem is to build a separate lookup table and keep it up-to-date. In the above case, you may build a separate username->user_id relationship table. Once you've informed V3 of this table, it will know what to do with a query like 'select * from user where username=:value'. You can also configure V3 to keep this table up-to-date as you insert or delete data. In other words, the application can be completely agnostic of this table's existence.
```
Comment (erzel) [inline doc/VTGateV3Features.md]: I like the example here, it makes it easier to understand.

File: doc/VTGateV3Features.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+* Joins that can be served by sending the query to multiple shards, and trivially combined to form the final result.
+* Cross-shard joins that can be served through a simple nested lookup.
+* Sorting that can be trivially merged from the results of multiple shards (merge-sort).
+* Aggregations (and grouping) that can be trivially combined from multiple shards.
+* A combination of the above constructs as long as the results remain trivially combinable.
+
+SQL is a very powerful language. You can build queries that can result in large amount of work and memory consumption involving big intermediate results. Such constructs where the scope of work is open-ended will not be immediately supported. In such cases, it's recommended that you use map-reduce techniques for which there is a separate API. *We'll consider building such on-the-fly map-reducers in the future.*
+
+## The vschema editor
+The above features require metadata like sharding key, cross-shard indexes, etc. to be configured and stored in some place. This is known as the vschema. *We'll be building a vschema editor and wizard that will allow you to easily view and modify this metadata.*
```
Comment (erzel) [inline doc/VTGateV3Features.md]: I'd rephrase as: "metadata like configuration of sharding key and cross-shard indexes" to make it clear that it's the configuration information that is being stored in the vschema not the entities themselves.

File: doc/VTGateV3Features.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
+# VTGate V3 features
+
+## Overview
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: As very first sentence, I would like to see here a definition of vtgate first. This will give a better introduction.

For me, vtgate is a proxy which contains the routing logic. You can also mention here that it's basically a fat client, but provided as a service. Over time, we moved away from a fat client, which had to be incorporated into the application, to the proxy model. This way, the routing logic is more decoupled from the application and easier to maintain separately.

File: doc/VTGateV3Features.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+With V3, the app does not need to specify any routing info. It just sends the query to VTGate as if it's a single database. Apart from simplifying the API, there are some additional benefits:
+* Database compliant drivers can be built for each client language. This will also allow for integration with third party tools that work with such drivers.
+* V3 can aspire to satisfy the full SQL syntax. This means that it will be able to perform cross-shard joins, aggregations and sorting.
+* Easier migration: an application that was written to use a single database can be trivially changed to use Vitess, and then the database can be scaled from underneath without changing much of the app.
+
+## Feature set
+The [V3 design](https://github.com/youtube/vitess/blob/master/doc/VTGateV3.md) is quite elaborate. If necessary, it will allow you to plug in custom indexes and sharding schemes. However, it comes equipped with some pre-cooked recipes that satisfy the immediate needs of the real-world:
+
+### Knowing where tables are
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: this section should probably to go the end of the feature set section? Most people are interested in sharding horizontally first and not vertically? Therefore, I would talk about horizontal sharding features first and then vertical sharding.

File: doc/VTGateV3Features.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+### Sharding schemes
+At its core, vitess uses range-based sharding, where the sharding column is typically a number or a varbinary. However, allowing data to be accessed only by the sharding key limits the flexibility of an application. V3 comes with a set of new indexing schemes that are built on top of range-based sharding.
+
+#### Basic sharding key
+If the application already has a well-distributed sharding key, you just have to tell VTGate what those keys are for each table. VTGate will correctly route your queries based on input values or the WHERE clause.
+
+#### Hashed sharding key
+If the application's sharding key is a monotonically increasing number, then you may not get well-balanced shards. In such cases, you can ask V3 to route queries based on the hash of the main sharding key.
+
+Vitess's filtered replication currently requires that the hash value be physically present as a column in each table. To satisfy this need, you still need to create a column to store this hash value. However, V3 will take care of populating this on your behalf.
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: Maybe put this in parentheses?

You could also slightly reorder the first sentence:

Currently, it is still required to store the hash value in an extra column in addition to the main sharding key because of Vitess' filtered replication feature. V3 will take care of populating the hash column on your behalf.

*The extra hash column will soon be no longer necessary when filtered replication is changed to perform the same hashing.*

File: doc/VTGateV3Features.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+*This restriction will soon be removed once filtered replication is changed to also perform the same hashing.*
+
+#### Auto-increment columns
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: these sections and the following should maybe go into their own section and not in "Sharding schemes" ? More like "Advanced Features" or something even more concrete? I don't know :)

File: doc/VTGateV3Features.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+#### Cross-shard indexes
+As your application evolves, you'll invariably find yourself wanting to fetch rows based on columns other than the main sharding key. For example, if you've sharded your database by user id, you may still want to be able find users by their username. If you only had the sharding key, such queries can only be answered by sending it to all shards. This could become very expensive as the number of shards grow.
+
+The typical strategy to address this problem is to build a separate lookup table and keep it up-to-date. In the above case, you may build a separate username->user_id relationship table. Once you've informed V3 of this table, it will know what to do with a query like 'select * from user where username=:value'. You can also configure V3 to keep this table up-to-date as you insert or delete data. In other words, the application can be completely agnostic of this table's existence.
+
+*Workflows can be developed to build these indexes on-the-fly.*
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: this isn't very concrete. Be more specific here or omit it?

File: doc/VTGateV3Features.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+#### Non-unique indexes
+Cross-shard indexes need not be unique. It is possible that rows may exist in multiple shards for a given where clause. V3 allows you to specify indexes as unique or non-unique, and accordingly enforces such constraints during changes.
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: s/need not be/don't have to be/

sounds better?

Comment (michael-berlin) [inline doc/VTGateV3Features.md]: not clear from the text what the implications are. Will this feature require an extra mapping table as well?

File: doc/VTGateV3Features.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+#### Shared indexes
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: reorder and put this section directly after "Cross-shard indexes" because this section is basically an extension of "Cross-shard indexes".

File: doc/VTGateV3Features.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+There are situations where multiple tables share the same foreign key. A typical use case is a situation where there is a customer table, an order table and an order_detail table. The order table would have a customer_id column. In order to efficiently access all orders of a customer, it would be beneficial to shard this table by customer_id. This will co-locate order rows with their corresponding customer row.
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: Use three backticks (example: ```TEXT```) to write the table names and column names as code. This way they are easier to distinguish and visually parse from the rest of the words in the sentence.

File: doc/VTGateV3Features.md (hunk 4 above)
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: I would prefer if you use the customer_id and order_id example here as well.

This way, the "Shared indexes" section can reuse the example and extend it.

File: doc/VTGateV3Features.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+The order table would also need an order_id column. As mentioned above, you can create an order_id->customer_id cross-shard index for this table. This will allow you to efficiently access orders by their order_id.
+
+In the case of an order_detail table, it may only need an order_id foreign key. Since this foreign key means the same thing as the order_id in order, creating a cross-shard index for it will result in a duplication of the order_id->customer_id index. In such situations, V3 allows you to just reuse the existing index for the order_detail table also. This saves disk space and also reduces the overall write load.
+
+### Consistency
+Once you add multiple indexes to tables, it's possible that the application could make inconsistent requests. V3 makes sure that none of the specified constraints are broken. For example, if a table had both a basic sharding key and a hashed sharding key, it will enforce the rule that the hash of the basic sharding key matches that of the hashed sharding key.
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: > For example, if a table had both a basic sharding key and a hashed sharding key

I don't think this makes sense. You're talking about the case here where we have two columns in the table e.g. "user_id" (basic sharding key) and "keyspace_id" (hashed sharding key)? Earlier you said you want to get rid of that anyway. This would make this feature redundant. Can you maybe give a better example?

Comment (michael-berlin) [inline doc/VTGateV3Features.md]: In addition to that, would it make sense to list all use cases? Otherwise this section may be too vague?

File: doc/VTGateV3Features.md [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+Some of the changes require updates to be performed across multiple databases. For example, inserting a row into a table that has a cross-shard key requires an additional row to be inserted into the lookup table. This results in distributed transactions. Currently, this is a best effort update. It is possible that partial commits happen if databases fail in the middle of a distributed commit.
+
+*There is a plan to support 2PC transactions to overcome this limitation.*
+
+### Query diversity
+V3 does not support the full SQL feature set. The current implementation supports simple queries:
+* Single table DML statements: This is a vitess-wide restriction where you can affect only one table and one sharding key per statement. *This restriction may be removed in the future.*
+* Single table SELECT statements:
+  * All constructs allowed if the statement targets only a single sharding key
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: s/allowed/are allowed/

File: doc/VTGateV3Features.md [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  * Aggregation and sorting not allowed if the statement targets more than one sharding key. Selects are allowed to target multiple sharding keys as long as the results from individual shards can be simply combined together to form the final result.
+
+Work is underway to support the following additional constructs:
+* Joins that can be served by sending the query to a single shard.
+* Joins that can be served by sending the query to multiple shards, and trivially combined to form the final result.
+* Cross-shard joins that can be served through a simple nested lookup.
```
Comment (michael-berlin) [inline doc/VTGateV3Features.md]: What's the difference between this and the previous item? I don't understand it. Maybe reformulate?

**Comment by michael-berlin** [comment]
I've read it and left some comments. Feel free to realize them.

File: doc/VTGateV3Features.md (hunk 9 above)
Comment (sougou) [inline doc/VTGateV3Features.md]: I think these should be listed here, because this is a features doc. If we were to build a user guide, these topics would belong to the advanced section. Hopefully, if the UI is intuitive enough, we can avoid building the user guide.

**Comment by sougou** [comment]
@michael-berlin PTAL.
I believe I've addressed all review comments. I feel like the paragraph hierarchy could be improved. Open to suggestions there.

**Comment by enisoc** [comment]
This is going to break some links when it's published. I'll update the links and then run publish.

**Comment by enisoc** [comment]
Published.
